File: packages/core/src/sinks.ts
/**
 * Default audit sinks for @bounded/core.
 *
 * Conforms to Bounded RFC v0 §7.3 and §9.4 (sink failure MUST NOT fail
 * evaluation — the engine catches; sinks SHOULD still try not to throw).
 *
 * Reference sinks shipped here:
 *   - ConsoleSink — stdout NDJSON (one JSON line per event)
 *   - NoOpSink    — drops events (useful for tests + emit_policy='off' scenarios)
 *   - MultiSink   — fan-out to N underlying sinks
 *
 * Not shipped here (future / separate packages):
 *   - FileSink    — rotating NDJSON file (needs fs + path conventions)
 *   - HttpSink    — POST to webhook with retry (needs retry strategy)
 *   - SqlAlchemySink / PrismaSink / DrizzleSink — storage adapter packages
 */

import type { IAuditEvent, IAuditSink } from './types.js';

/** NDJSON-to-stdout sink. Synchronous writes. */
export class ConsoleSink implements IAuditSink {
  constructor(
    private readonly out: (line: string) => void = (s) => {
      process.stdout.write(s);
    },
  ) {}

  emit(event: IAuditEvent): void {
    this.out(JSON.stringify(event) + '\n');
  }
}

/** Discards every event. Returns resolved flush. */
export class NoOpSink implements IAuditSink {
  emit(_event: IAuditEvent): void {
    // intentional
  }
  async flush(): Promise<void> {}
}

/**
 * Fan-out to multiple sinks. Failures in one child sink do NOT stop delivery
 * to the others — each sink is emitted to independently.
 */
export class MultiSink implements IAuditSink {
  private readonly children: readonly IAuditSink[];

  constructor(children: readonly IAuditSink[]) {
    this.children = [...children];
  }

  async emit(event: IAuditEvent): Promise<void> {
    const errors: unknown[] = [];
    await Promise.all(
      this.children.map(async (child) => {
        try {
          await child.emit(event);
        } catch (err) {
          errors.push(err);
        }
      }),
    );
    // If every child failed, surface the first error so the engine-level
    // try/catch (RFC §9.4) still swallows — but caller / tests can detect
    // total failure if they want.
    if (errors.length > 0 && errors.length === this.children.length) {
      throw errors[0];
    }
  }

  async flush(): Promise<void> {
    await Promise.all(this.children.map((c) => c.flush?.()));
  }
}

File: python/src/themis/sinks.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Adam Campbell
"""Reference audit sinks (RFC v0 § 7.3, § 9.4)."""
from __future__ import annotations

import json
import sys
from typing import Callable, Optional, Sequence

from .types import AuditEvent


class ConsoleSink:
    """NDJSON to stdout — one JSON line per event."""

    def __init__(self, out: Optional[Callable[[str], None]] = None) -> None:
        self._out = out or (lambda s: sys.stdout.write(s))

    def emit(self, event: AuditEvent) -> None:
        self._out(json.dumps(event.to_dict(), default=str) + "\n")


class NoOpSink:
    def emit(self, event: AuditEvent) -> None:  # noqa: D401 — intentional
        pass

    def flush(self) -> None:
        pass


class MemorySink:
    """Collects events in memory — for tests and short-lived processes."""

    def __init__(self) -> None:
        self.events: list[AuditEvent] = []

    def emit(self, event: AuditEvent) -> None:
        self.events.append(event)


class MultiSink:
    """Fan-out. One child failing never stops delivery to the others; if EVERY
    child fails the first error is raised so the engine's § 9.4 catch sees it."""

    def __init__(self, children: Sequence) -> None:
        self._children = list(children)

    def emit(self, event: AuditEvent) -> None:
        errors: list[BaseException] = []
        for child in self._children:
            try:
                child.emit(event)
            except Exception as e:  # noqa: BLE001
                errors.append(e)
        if errors and len(errors) == len(self._children):
            raise errors[0]

    def flush(self) -> None:
        for child in self._children:
            flush = getattr(child, "flush", None)
            if callable(flush):
                flush()


__all__ = ["ConsoleSink", "NoOpSink", "MemorySink", "MultiSink"]
